### DISCUSSION TITLE
How to handle date formatting in forms

### DISCUSSION THREAD

**Discussion by wurst-hans** [phalcon/cphalcon#15498] [Q&A]
When working with forms and models, what is the best practise for handling different locales?

In database, all dates are stored in format `yyyy-mm-dd`. Now, I want to write an application for german locale and date format is `dd.mm.yyyy` for Germany. So between UI (input field) and form/model the date format has to be converted in both directions.

How can I transform database date to german locale before rendering input field? Additionally, what is best practise to handle a date in german format that is submitted to form by user when it comes to form validation and storing in database?

**Comment by wurst-hans** [comment]
First part is solved. I extended `Form` class and have overwritten `getValue()`. I intentionally did *not* use `getCustomValue()`. So, this is my code to handle conversion of entity value for rendering input field:

```
public function getValue(string $name)
{
    if ($this->entity instanceof Model) {
        $metadata = $this->modelsMetadata;
        $types = $metadata->getDataTypes($this->entity);
        if (
            ($types[$name] ?? null) === Column::TYPE_DATE &&
            !is_null($this->entity->$name)
        ) {
            return $this->locale->formatDate($this->entity->$name);
        }
    }
    return parent::getValue($name);
}
```

Please notice, that `$this->locale->formatDate()` is a custom service in my project, which uses `IntlDateFormatter` for rendering date in target date format.

I have to solve second part now, i.e. when form is submitted back to application, then german date format has to be converted to `yyyy-mm-dd` format.

**Comment by Jeckerson** [comment]
Hello, try use `Carbon` library.
